### ISSUE TITLE
Custom naming for image file

### ISSUE THREAD

**Issue by gmoro** [OSInside/kiwi#1943]
Currently there is no way of customizing the naming of the output image.
For our current practical example on JeOS (that is currently being renamed to Minimal):
The definition would be
`<image  name="SLES15-SP3-JeOS" .../>`
Building the OpenStack profile defined as:
`<profile name="OpenStack-Cloud" description="JeOS for OpenStack Cloud" arch="x86_64"/>`
And having the profile configured with
`<preferences profiles="OpenStack-Cloud">
        <version>15.3</version>`
we would get an image named

`SLES15-SP1-JeOS.x86_64-15.1-OpenStack-Cloud-Build.qcow2`

Canonically meaning IMAGE_NAME.ARCH-PROFILE_VERSION-PROFILE_NAME-BUILD_VERSION.FORMAT

in the rename of JeOS to Minimal we face one specific challenge, we want to separate the Minimal-VM from the Minimal-Image, meaning, the VM <type>image="vmx" from the  image="oem", but profile preferences don't actually help much here.

Ideally we would like to be able to achieve

`SLES15-SP4-Minimal-VM.x86_64-OpenStack-Cloud-GM.qcow2
SLES15-SP4-Minimal-Image.aarch64-RaspberryPi-GM.raw.xz`

meaning canonically something like 
IMAGE_NAME.BUILD_TYPE.ARCH-PROFILE_NAME-BUILD_VERSION.FORMAT

Basically dropping the profile version that is currently not really useful given that the version information is already on the NAME tag (15-SP4) and having a way to differentiate the BUILD_TYPE between vmx and oem

In a more complete solution would be nice to be able to have the actual version (15-SP4) split from the name as well. Maybe even splitting or creating more fields.

Of course I understand is not a simple change, but having an extra mechanism that can define canonical naming based on existing fields and maybe adding a few more fields related to naming specifically might make the naming more robust, including support multiple naming schemes of different distributions.

What are you guys suggestions and ideas regarding that?

**Comment by vmoutoussamy** [comment]
For ref this is link to jsc#SLE-20354 the epic for SLES 15 SP4. Please note that on the pure SLE side, the autobuild team has script in place that could do the rename for us (in the comments of the Epic). However there is no such scripts from the openSUSE side thus we are not willing to create a workaround script only for openSUSE JeOS images in the openSUSE release process.
Last but not least we believe that the autobuild script is a workaround because of the kiwi limitation exposed by Gui in the description thus fixing kiwi will allow us to proceed with the rename for openSUSE and make the SLE autobuild script obsolete for the rename!

**Comment by schaefi** [comment]
And of course there are issues. Currently obs decides about the contents of the --id option. In the example above this is

```
--id BIOS-Build11.2
```

no matter what format we define this id setting has to continue to be respected in the same way as before, otherwise we break what obs expects.

I'm not aware of a way to prevent this setting by obs

**Comment by schaefi** [comment]
Another issue is time. I could spent almost all of my day to work on kiwi but actually there are other priorities on the job in the pubcloud team. So in other words how urgent is this ? I doubt that @davidcassany has time to work on it either since he is booked with the rancher team priorities. If it can sit and wait for some time we don't have an issue

**Comment by gmoro** [comment]
> 
>     1. We add a new attribute to the kiwi schema
>        ```
>        <image ... bundle_format="%NAME-%ARCH..."/>
>        ```
> 
>     2. We extend kiwi to embed the bundle_format in the .result file. This is a binary format which serializes the result object from the build into a file. The file is read by the `kiwi result bundle` call
> 
>     3. We extend the result bundle code in kiwi to handle the format and create the files in this format
> 
> 
> In a very first step we need to declare which format options we need and makes sense. If no format options are specified the current naming applies, that way this can be done backward compatible
> 
> Thoughts on this plan ?
That would improve a lot the situation but we would still have the problem of having to differentiate the file name based on the profile preferences.

It is specially tricky because it would be a "if" based on the type="vmx" or we would to have to include a bundle_format at the profile preference level as well, taking priority

**Comment by schaefi** [comment]
>That would improve a lot the situation but we would still have the problem of having to differentiate the file name based on the profile preferences.

yeah good point. In this case we could also move the bundle_format attribute into the ```<profiles>``` section. However, this section is optional because profiles are not mandatory.

It probably would fit best into ```<type>```

```xml
<type image="..." ... bundle_format="..."/>

<type image="..." ... bundle_format="..." profiles="..."/>
```

That way different bundle formats per profile can be specified.

Still leaves the problem that obs passes --id which the bundler could ignore if a bundle_format is available

**Comment by gmoro** [comment]
> It probably would fit best into `<type>`
> 
> ```
> <type image="..." ... bundle_format="..."/>
> 
> <type image="..." ... bundle_format="..." profiles="..."/>
> ```
> 
> That way different bundle formats per profile can be specified.

Seems better.
In the case JeOS.kiwi goes like
<preferences profiles="kvm-and-xen">
 <type bundle_format="...." />

That's the idea?

> 
> Still leaves the problem that obs passes --id which the bundler could ignore if a bundle_format is available

I guess this is not such a big problem, maybe the default should be to include --id anyway, to avoid breaking OBS , and leave as it is, adding as part of the version when present (always, because version is mandatory in at least one preference)
`                        image_version + '-' + self.command_args['--id']
`
Also would be nice to be able to ignore the <version> tag, as we just decided to reset our value to 1.0.0 because the "15.4" version tag we used before is more confusing then helping, and in the case of Tumbleweed does not even make sense, and the only real information about version is the Build ID, as we still don't have a "VERSION_ID="20210929" (as it goes into /etc/os-release) for the JeOS images (that's a separate thing).

In the case of our image OBS send --id=kvm-and-xen-Build43.194" which is profile_name+build_id, I noticed that even locally with osc we get --id=profile_name+"Build" with no number.

Or maybe %ID should be one of the fields available for the bundle_format and ignored if not there? I guess this is what you propose? This way the default behaviour would still be used when no bundle_format is not present and otherwise it can be moved around in the name if you want.

**Comment by gmoro** [comment]
In a separate note, we were trying to come up with a really canonical way of defining what an image would need to express most possibilities

NAME=openSUSE-Tumbleweed
IMAGE_VERSION=20211006
IMAGE_TYPE=VM (image="vmx" on the type tag)
ARCH=x86_64
BUILD_ID=kvm-and-xen-Build12.3
OR even more fine grained
BUILD_FLAVOR=kvm-and-xen (in kiwi really is a profile_name)
BUILD_VERSION=Build12.3 (Not really available to kiwi)

Or for SLE:

NAME=SLES
IMAGE_VERSION=15-SP4
IMAGE_TYPE=Image (image="oem" on the type tag)
ARCH=x86_64
BUILD_ID=kvm-and-xen-Build12.3

Just so you can understand what fields could be useful for us, just having a IMAGE_TYPE=Image or IMAGE_TYPE=VM field on the profile would be enough for us to use to construct the bundle_format using %NAME%IMAGE_TYPE%ARCH, but I think this is a less clean way, although we would have no need to include a floating string with "VM" on the bundle_format="..."
